### ISSUE TITLE
WRONG code  SLP vectorizer? InstCombine?

### ISSUE THREAD

**Issue by JonPsson1** [llvm/llvm-project#75995]
```
int printf(const char *, ...);

unsigned char a = 255;
unsigned int b = 0;

unsigned short fun(unsigned char g) { return  g - 1; }

int main() {
  for (short c = 0; c < 8; c++)
    b |= fun(--a);

  printf("%X\n", b);
}

```
fun() is doing arithmetic with 8 bits, and extending the result to 16 bits. 'b' ORs itself with this result, which could never be more than 8 bits. Yet, I see:

```
clang -target s390x-linux-gnu -march=z14 -O3 wrong1.i -o ./a.out; ./a.out
FFFF

clang -target s390x-linux-gnu -march=z14 -O3 wrong1.i -mllvm -slp-vectorize-hor=false -o ./a.out; ./a.out;
FF
```
It seems SLP vectorizer is generating an @llvm.vector.reduce.or.v8i16 which is later removed by InstCombine...

**Comment by topperc** [comment]
-opt-bisect-limit points to InstCombine

**Comment by topperc** [comment]
I think it is an SLP bug. SLP doesn't drop nsw/nuw flags when it turns i32 scalar adds into an i16 vector add.

@alexey-bataev @nikic

**Comment by topperc** [comment]
Code before SLP

```
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load i8, ptr @a, align 1, !tbaa !6
  %2 = load i32, ptr @b, align 4, !tbaa !9
  %3 = add i8 %1, -1
  %4 = zext i8 %3 to i32
  %5 = add nuw nsw i32 %4, 65535
  %6 = add i8 %1, -2
  %7 = zext i8 %6 to i32
  %8 = add nuw nsw i32 %7, 65535
  %9 = or i32 %5, %8
  %10 = add i8 %1, -3
  %11 = zext i8 %10 to i32
  %12 = add nuw nsw i32 %11, 65535
  %13 = or i32 %9, %12
  %14 = add i8 %1, -4
  %15 = zext i8 %14 to i32
  %16 = add nuw nsw i32 %15, 65535
  %17 = or i32 %13, %16
  %18 = add i8 %1, -5
  %19 = zext i8 %18 to i32
  %20 = add nuw nsw i32 %19, 65535
  %21 = or i32 %17, %20
  %22 = add i8 %1, -6
  %23 = zext i8 %22 to i32
  %24 = add nuw nsw i32 %23, 65535
  %25 = or i32 %21, %24
  %26 = add i8 %1, -7
  %27 = zext i8 %26 to i32
  %28 = add nuw nsw i32 %27, 65535
  %29 = or i32 %25, %28
  %30 = add i8 %1, -8
  %31 = zext i8 %30 to i32
  %32 = add nuw nsw i32 %31, 65535
  %33 = or i32 %29, %32
  %34 = and i32 %33, 65535
  %35 = or i32 %34, %2
  store i8 %30, ptr @a, align 1, !tbaa !6
  store i32 %35, ptr @b, align 4, !tbaa !9
  %36 = tail call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef %35)
  ret i32 0
}
```

Code after SLP

```
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load i8, ptr @a, align 1, !tbaa !6
  %2 = load i32, ptr @b, align 4, !tbaa !9
  %3 = insertelement <8 x i8> poison, i8 %1, i32 0
  %4 = shufflevector <8 x i8> %3, <8 x i8> poison, <8 x i32> zeroinitializer
  %5 = add <8 x i8> %4, <i8 -2, i8 -1, i8 -3, i8 -4, i8 -5, i8 -6, i8 -7, i8 -8>
  %6 = zext <8 x i8> %5 to <8 x i16>
  %7 = add nuw nsw <8 x i16> %6, <i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1>
  %8 = call i16 @llvm.vector.reduce.or.v8i16(<8 x i16> %7)
  %9 = zext i16 %8 to i32
  %10 = and i32 %9, 65535
  %11 = or i32 %10, %2
  %12 = extractelement <8 x i8> %5, i32 7
  store i8 %12, ptr @a, align 1, !tbaa !6
  store i32 %11, ptr @b, align 4, !tbaa !9
  %13 = tail call i32 (ptr, ...) @printf(ptr noundef nonnull dereferenceable(1) @.str, i32 noundef %11)
  ret i32 0
}
```

**Comment by alexey-bataev** [comment]
Yes, need to drop wrapping flags, the patch is almost ready

**Comment by alexey-bataev** [comment]
Fixed in a13148a8808d1cc50c97cb2c1411694840a5ddc7
